### ISSUE TITLE
Make sure Universal works with new UI Router

### ISSUE THREAD

**Issue by jeffwhelpley** [angular/universal#184]
Test out Angular Universal with the new UI Router and fix any bugs that come up:

https://github.com/angular-ui/ui-router/tree/feature-1.0

**Comment by christopherthielen** [comment]
`feature-1.0` branch is not ng2... I've got a plunker with the current ng2 prototype though: http://plnkr.co/edit/PF69BVa1WnUYZKkURgMm

**Comment by jeffwhelpley** [comment]
Gotcha. Thanks for the info, @christopherthielen. I am going to leave this open and it would be cool if you let us know when the ng2 adapter is ready. We can make sure it works with Angular Universal.

**Comment by jeffwhelpley** [comment]
@christopherthielen what is the latest on this. I heard a rumor that the UI Router may be ready in a month or so to work with Angular 2. On a side note, in addition to getting the new UI Router to work with Universal, I am also interested in having you on Angular Air once it is ready to talk about it.

**Comment by christopherthielen** [comment]
I think "about a month" is still a reasonable estimate.  It will probably only support hash bang urls at that time, but otherwise be relatively functional

**Comment by jeffwhelpley** [comment]
ok, just let us know. thanks!

**Comment by christopherthielen** [comment]
circling back around on this:

UI-Router [Version 1.0.0-alpha.3](https://github.com/angular-ui/ui-router/releases/tag/1.0.0-alpha.3) has initial ng2 support.

**Comment by jeffwhelpley** [comment]
@christopherthielen does it support pushstate? I think you mentioned that at first it was only going to do hash urls. We sort of need pushstate in order to integrate with universal.

**Comment by christopherthielen** [comment]
UI-Router ng2 doesn't support pushstate yet. Can you explain what we need to do to be compatible with universal?  At what level do we need to tie into push state?

I see that component router uses multiple levels of abstraction:

First is the url writing strategy `LocationStrategy`... for instance: `HashLocationStrategy`. 
The strategy delegates to a `PlatformLocation`, i.e. the `BrowserPlatformLocation` (still in `angular/router` code so far)
The `BrowserPlatformLocation` delegates to the `DOM` class from `angular2/src/platform/dom/dom_adapter`

Would it be best to code to `DOM.getHistory()`, `DOM.getLocation()`, `DOM.getBaseHref()`, like `BrowserPlatformLocation` from component router is doing?

**Comment by jeffwhelpley** [comment]
Well, the problem is that browsers don't send hash URLs to the server. So, there would be no way to know on the server what route is being requested. Do you have plan to support pushstate in the future?

**Comment by christopherthielen** [comment]
> Do you have plan to support pushstate in the future?

Oh yes, definitely.  

I'm just hoping to understand if I need to hook into (and/or provide) any specific abstractions for ui-router to work with universal.  For instance, if I simply used window.location and window.history, would that be a problem?

**Comment by jeffwhelpley** [comment]
Well, we encourage devs and 3rd party libs to use the Angular 2 core DOM abstractions, but I believe you wanted to create UI router without any dependency on Angular 2, right? Assuming that is the case, it will still work, but we just need to include a patch for Universal which essentially mocks out certain window/browser/DOM objects like location and history on the server side. Outside that, the only other complication would be ensuring Universal knows when to return a response to the client from the server when using UI Router. In the older version of UI Router and Angular 1, I know this would be as simple as listening for $stateChangeSuccess (not sure if similar thing in latest version), but that assumes all async is in the UI Router resolve. 

Anyways, let's not digress too much here. I will ping the Universal community and see if I can get someone who is able to work with you on this. I would love to have a prototype before ng-conf, but no big deal if it doesn't work out.

**Comment by christopherthielen** [comment]
> I believe you wanted to create UI router without any dependency on Angular 2, right?

The UI-Router core, yes, but the ng2 adapter for ui-router-ng2 ties directly to angular 2 code.

> Well, we encourage devs and 3rd party libs to use the Angular 2 core DOM abstractions

By core DOM abstractions, are we talking about the `DOM` object?
```
import {DOM} from "angular2/src/platform/dom/dom_adapter"
```

**Comment by christopherthielen** [comment]
I think push state is ready.  I've got the code in this branch at the moment: https://github.com/angular-ui/ui-router/tree/ocombe-html5mode

**Comment by jeffwhelpley** [comment]
OK, I am adding this back to the active list and we will talk about it at the next community meeting.

**Comment by christopherthielen** [comment]
Just released alpha.5 with pushstate support using `LocationProvider`

**Comment by jeffwhelpley** [comment]
wu-hu! I am meeting with the Universal team today and will see if I can get someone to help trying out an integration between the latest UI Router and Angular Universal. I will let you know if I have any questions.

**Comment by laco0416** [comment]
any updates on this?

**Comment by jeffwhelpley** [comment]
Hi @laco0416. This is on our list, but is a lower priority. The critical path now is getting to production-ready status. That said, we would love if someone from the community had time to take this one. If you or anyone else is interested let me know. Patrick and I as well as Chris from the UI Router project can help provide guidance.

**Comment by lngr** [comment]
Shall the new router the only working router, or must there additional support for the deprecated router?

**Comment by jeffwhelpley** [comment]
@Ingr getting the new router working with Universal is a separate effort than this issue, but to answer your question, are moving in the direction of trying to support the new router and the UI router. The deprecated router, is...well, deprecated.

**Comment by gdi2290** [comment]
the new ui router should work so long as zone.js is able to catch it as that's how we're handling all of the other routers
